plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -513,6 +515,35 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
         return getTableMetadata(session, ((IcebergTableHandle) table).getSchemaTableName());
     }
 
+    @Override
+    public ConnectorTableSchema getTableSchema(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        Schema schema = getSchema(session, table);
+        List<ColumnMetadata> columns = getColumnMetadatas(schema);
+
+        return new ConnectorTableSchema(
+                table.getSchemaTableName(),
+                columns.stream()
+                        .map(ColumnMetadata::getColumnSchema)
+                        .collect(toUnmodifiableList()));
+    }
+
+    private Schema getSchema(ConnectorSession session, IcebergTableHandle table)
+    {
+        Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
+        if (table.getSnapshotId().isEmpty() || table.getSnapshotId().get() == icebergTable.currentSnapshot().snapshotId()) {
```

Comment:
`table.getSnapshotId()` is always filled even when not doing snapshot queries because it is the way to see whether the table has or not data in `IcebergSplitManager` or `TableStatisticsMaker`.

With this value always filled, there is currently no way to know whether a specific snapshot of the table is being queried.

Reply:
This is desired behavior (explained here: https://github.com/trinodb/trino/pull/12786#discussion_r894648522)

what you discovered is that snapshot-id doesn't describe table's state fully, in case there was ADD COLUMN (and no further INSERT yet). 
We just need to capture what identifies the table state. I assume this is `(snapshot-id, schema-id)` pair. The table handle should therefore carry both.

Reply:
I've modified the logic of `getTableHandle()` method so that the correct `tableSchemaJson` and `partitionSpecJson` are transported over the `IcebergTableHandle`.
Thank you @findepi  for hinting me towards this path.